pkg/costattribution/sample_tracker.go

```diff
@@ -0,0 +1,302 @@
+	if st.overflowSince.Load() > 0 {
+		st.overflowCounter.receivedSample.Add(receivedSampleIncrement)
+		if discardedSampleIncrement > 0 && reason != nil {
+			st.overflowCounter.totalDiscarded.Add(discardedSampleIncrement)
+		}
+		st.observedMtx.RUnlock()
+		return
+	}
+
+	o, known := st.observed[key]
+	if known && st.overflowSince.Load() == 0 {
+		o.lastUpdate.Store(ts.Unix())
+		o.receivedSample.Add(receivedSampleIncrement)
+		if discardedSampleIncrement > 0 && reason != nil {
+			o.discardedSampleMtx.Lock()
+			if _, ok := o.discardedSample[*reason]; ok {
+				o.discardedSample[*reason].Add(discardedSampleIncrement)
+			} else {
+				o.discardedSample[*reason] = atomic.NewFloat64(discardedSampleIncrement)
+			}
+			o.discardedSampleMtx.Unlock()
+		}
+		st.observedMtx.RUnlock()
+		return
+	}
+	st.observedMtx.RUnlock()
+
+	// If it is not known, we take the write lock, but still check whether the key is added in the meantime
+	st.observedMtx.Lock()
+	defer st.observedMtx.Unlock()
+	// If not in overflow, we update the observation if it exists, otherwise we check if create a new observation would exceed the max cardinality
+	// if it does, we set the overflowSince
+	if st.overflowSince.Load() == 0 {
+		o, known = st.observed[key]
+		if known {
+			o.lastUpdate.Store(ts.Unix())
+			o.receivedSample.Add(receivedSampleIncrement)
+			if discardedSampleIncrement > 0 && reason != nil {
+				o.discardedSampleMtx.Lock()
+				if _, ok := o.discardedSample[*reason]; ok {
+					o.discardedSample[*reason].Add(discardedSampleIncrement)
+				} else {
+					o.discardedSample[*reason] = atomic.NewFloat64(discardedSampleIncrement)
+				}
+				o.discardedSampleMtx.Unlock()
+			}
+			return
+		}
+		// if it is not known, we need to check if the max cardinality is exceeded
+		if len(st.observed) >= st.maxCardinality {
+			st.overflowSince.Store(ts.Unix())
+		}
+	}
+
+	// if overflowSince is set, we only update the overflow counter
+	if st.overflowSince.Load() > 0 {
+		st.overflowCounter.receivedSample.Add(receivedSampleIncrement)
+		if discardedSampleIncrement > 0 && reason != nil {
+			st.overflowCounter.totalDiscarded.Add(discardedSampleIncrement)
+		}
+		return
+	}
+
+	// create a new observation
+	st.observed[key] = &observation{
+		lastUpdate:         *atomic.NewInt64(ts.Unix()),
+		discardedSample:    make(map[string]*atomic.Float64),
+		receivedSample:     *atomic.NewFloat64(receivedSampleIncrement),
+		discardedSampleMtx: sync.Mutex{},
+	}
+
+	if discardedSampleIncrement > 0 && reason != nil {
+		st.observed[key].discardedSampleMtx.Lock()
+		st.observed[key].discardedSample[*reason] = atomic.NewFloat64(discardedSampleIncrement)
+		st.observed[key].discardedSampleMtx.Unlock()
+	}
+}
+
+func (st *SampleTracker) recoveredFromOverflow(deadline time.Time) bool {
+	st.observedMtx.RLock()
+	if st.overflowSince.Load() > 0 && time.Unix(st.overflowSince.Load(), 0).Add(st.cooldownDuration).Before(deadline) {
+		if len(st.observed) <= st.maxCardinality {
```

Comment:
Previously we checked:
```go
		// if it is not known, we need to check if the max cardinality is exceeded
		if len(st.observed) >= st.maxCardinality {
			st.overflowSince.Store(ts.Unix())
		}
```
So _equals_ meant we're in overflow, but now _equals_ means we've recovered from overflow. This is going to cause continous flapping there.

Reply:
changed the condition to strict less than maxcardinality in the mean time e5cda41a140083522dffcee0aad1cab3ff73d415